devito/interfaces.py

```diff
@@ -440,44 +445,126 @@ class DenseData(TensorData):
+                setattr(self.__class__, 'd%s4' % dim.name,
+                        property(dx4, 'Symbol for the fourth derivative '
+                                 'wrt the %s dimension' % dim.name))
+
+                for dim2 in self.space_dimensions:
+                    # First cross derivative
+                    dxy = partial(cross_derivative, order=self.space_order,
+                                  dims=(dim, dim2))
+                    setattr(self.__class__, 'd%s%s' % (dim.name, dim2.name),
+                            property(dxy, 'Symbol for the cross derivative '
+                                     'wrt the %s and %s dimension' %
+                                     (dim.name, dim2.name)))
+
+                    # Second cross derivative
+                    dx2y2 = partial(second_cross_derivative, dims=(dim, dim2),
+                                    order=self.space_order)
+                    setattr(self.__class__, 'd%s2%s2' % (dim.name, dim2.name),
+                            property(dx2y2, 'Symbol for the cross derivative '
+                                     'wrt the %s and %s dimension' %
+                                     (dim.name, dim2.name)))
+
     @classmethod
     def _indices(cls, **kwargs):
         """Return the default dimension indices for a given data shape
 
+        :param grid: :class:`Grid` that defines the spatial domain.
         :param dimensions: Optional, list of :class:`Dimension`
                            objects that defines data layout.
-        :param shape: Optional, shape of the spatial data to
-                      automatically infer dimension symbols.
         :return: Dimension indices used for each axis.
+
+        ..note::
+
+        Only one of :param grid: or :param dimensions: is required.
         """
+
+        grid = kwargs.get('grid', None)
         dimensions = kwargs.get('dimensions', None)
-        if dimensions is None:
-            # Infer dimensions from default and data shape
-            if 'shape' not in kwargs:
-                error("Creating symbolic data objects requries either"
-                      "a 'shape' or 'dimensions' argument")
+        if grid is None:
+            if dimensions is None:
+                error("Creating a Function object requries either "
+                      "a 'grid' or the 'dimensions' argument.")
                 raise ValueError("Unknown symbol dimensions or shape")
-            _indices = (x, y, z)
-            shape = kwargs.get('shape')
-            if len(shape) <= 3:
-                dimensions = _indices[:len(shape)]
-            else:
-                dimensions = [Dimension("x%d" % i) for i in range(1, len(shape) + 1)]
+        else:
+            if dimensions is not None:
+                warning("Creating Function with 'grid' and 'dimensions' "
+                        "argument; ignoring the 'dimensions' and using 'grid'.")
+            dimensions = grid.dimensions
```

Comment:
should be only line needed if `grid` created in `init` if missing.

Reply:
Ahhh, now I get you. However, I don't think that is a good idea, since, for example, a point source has dimensions `(d, npoints)`, which means I would now have to auto-create a grid for that dimension pair. As a grid is actually meant to represent a physical space, that makes very little sense though, which is why we need to keep the ability to explicitly specify a function with a `(shape, dimensions)` pair.